src/libaktualizr/primary/aktualizr.cc

```diff
@@ -32,10 +33,158 @@ Aktualizr::Aktualizr(Config &config) : config_(config) {
   uptane_client_->initialize();
 }
 
+void Aktualizr::FinalizeInstall() {
+  // TODO: is it too much?
+  FetchMetadata();
+
+  boost::filesystem::path reboot_flag = "/tmp/aktualizr_reboot_flag";
+  if (boost::filesystem::exists(reboot_flag)) {
+    boost::filesystem::remove(reboot_flag);
+    if (getppid() == 1) {  // if parent process id is 1, aktualizr runs under systemd
+      exit(0);             // aktualizr service runs with 'Restart=always' systemd option.
+      // Systemd will start aktualizr automatically if it exit.
+    } else {  // Aktualizr runs from terminal
+      LOG_INFO << "Aktualizr has been updated and requires restarting to run the new version.";
+    }
+  }
+}
+
+void Aktualizr::UptaneCycle() {
```

Comment:
There is a fair bit of redundancy between these three new functions. I think it's okay to use running_mode here to decide on actions. I'm not positive that combining these into one function is the right answer, but it might be.

Also, I'm loosely in favor of making the lambda functions "real" functions so they are a bit more obvious/readable/etc. Not a requirement, though, and I could be convinced otherwise.

Reply:
Clarification: I think lambdas are best for short things, but when it gets to be this long (more than a few lines), I prefer "proper" functions.

Reply:
Yes, I have a similar opinion about lambda actually but here the capture list is not trivial so I can either create a state object to contain them or have them as arguments and make a big `std::bind`. These are also ugly to some extent, but I'll try to reevaluate the options (maybe a struct wrapper). In their favor, you could argue that everything is at the same place.

About having three similar functions, I thought that the code would be more self-documenting and would avoid some spaghetti/goto effect. But maybe that's excessive and will actually complicate maintenance. I'll also try the other way.

Reply:
Follow your heart. I'm really not sure yet what is best here, and the first rule is what works. Anything that cleans up SotaUptaneClient is an improvement, although there is a limit to how much we should muck around with Aktualizr.